### PR TITLE
Feat-23 댓글 작성, 수정, 삭제 구현

### PR DESCRIPTION
**Pull request by EunjungOh123**

### CODE REVIEW

**Comment by naekang** [comment]
|File|Coverage [71.63%]|:x:|
|:-|:-:|:-:|
|[CommentController.java](https://github.com/Gaethering/Gaethering-Server/blob/61392c25eef0942ede2ea45f6a93f5b703af5d56/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fcontroller%2FCommentController.java)|100%|:green_apple:|
|[CommentServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/61392c25eef0942ede2ea45f6a93f5b703af5d56/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fservice%2FCommentServiceImpl.java)|100%|:green_apple:|
|[PostErrorCode.java](https://github.com/Gaethering/Gaethering-Server/blob/61392c25eef0942ede2ea45f6a93f5b703af5d56/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fexception%2FerrorCode%2FPostErrorCode.java)|100%|:green_apple:|
|[PostServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/61392c25eef0942ede2ea45f6a93f5b703af5d56/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fservice%2FPostServiceImpl.java)|82.47%|:green_apple:|
|[Category.java](https://github.com/Gaethering/Gaethering-Server/blob/61392c25eef0942ede2ea45f6a93f5b703af5d56/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fentity%2FCategory.java)|74.32%|:x:|
|[Comment.java](https://github.com/Gaethering/Gaethering-Server/blob/61392c25eef0942ede2ea45f6a93f5b703af5d56/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fentity%2FComment.java)|73.63%|:x:|
|[Post.java](https://github.com/Gaethering/Gaethering-Server/blob/61392c25eef0942ede2ea45f6a93f5b703af5d56/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fentity%2FPost.java)|68.02%|:x:|
|[Heart.java](https://github.com/Gaethering/Gaethering-Server/blob/61392c25eef0942ede2ea45f6a93f5b703af5d56/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fentity%2FHeart.java)|4.17%|:x:|
|[PostImage.java](https://github.com/Gaethering/Gaethering-Server/blob/61392c25eef0942ede2ea45f6a93f5b703af5d56/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fentity%2FPostImage.java)|3.61%|:x:|

|Total Project Coverage|64.25%|:x:|
|:-|:-:|:-:|

File: src/main/java/com/gaethering/gaetheringserver/domain/board/exception/errorCode/PostErrorCode.java [hunk 1]
```diff
@@ -10,7 +10,10 @@ public enum PostErrorCode implements ErrorCode {
 
     POST_NOT_EXIST("E801", "존재하지 않는 게시물입니다."),
     CATEGORY_NOT_EXIST("E802", "존재하지 않는 카테고리입니다."),
-    CANNOT_UPDATE_POST("E802", "게시물 수정 권한은 작성자에게 있습니다.");
+    COMMENT_NOT_EXIST("E803", "존재하지 않는 댓글입니다."),
+    CANNOT_UPDATE_COMMENT("E804", "댓글 수정 권한은 작성자에게 있습니다."),
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/exception/errorCode/PostErrorCode.java:14]: enum 명만 봤을 때 구체적으로 어떤 이유에서 나는 에러인지 모호한 것 같습니다!

File: src/main/java/com/gaethering/gaetheringserver/domain/board/exception/errorCode/PostErrorCode.java [hunk 2]
```diff
@@ -10,7 +10,10 @@ public enum PostErrorCode implements ErrorCode {
...
+    CANNOT_DELETE_COMMENT("E805", "댓글 삭제 권한은 작성자에게 있습니다."),
+    CANNOT_UPDATE_POST("E806", "게시물 수정 권한은 작성자에게 있습니다.");
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/exception/errorCode/PostErrorCode.java:16]: 위 두개도 위에 코멘트랑 비슷한 의견입니다

File: src/main/java/com/gaethering/gaetheringserver/domain/board/exception/FailDeleteCommentException.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.gaethering.gaetheringserver.domain.board.exception;
+
+import com.gaethering.gaetheringserver.domain.board.exception.errorCode.PostErrorCode;
+
+public class FailDeleteCommentException extends PostException {
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/exception/FailDeleteCommentException.java:5]: 예외 클래스명이 권한이 없다 라는 의미가 보이지 않는 것 같습니다. 의도가 명확하게 보이는 이름으로 하는게 어떨까요?

File: src/main/java/com/gaethering/gaetheringserver/domain/board/exception/FailUpdateCommentException.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.gaethering.gaetheringserver.domain.board.exception;
+
+import com.gaethering.gaetheringserver.domain.board.exception.errorCode.PostErrorCode;
+
+public class FailUpdateCommentException extends PostException{
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/exception/FailUpdateCommentException.java:5]: 이것도 삭제에 드렸던 코멘트 내용과 비슷한 의견입니다!

File: src/main/java/com/gaethering/gaetheringserver/domain/board/controller/CommentController.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import javax.validation.Valid;
+import java.security.Principal;
+
+@RestController
+@RequestMapping("/api/boards")
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/controller/CommentController.java:16]: /api 같은 부분은 yml을 이용해서 prefix 처리하는게 어떨까요? 저는 그런 방식을 사용했습니다!

File: src/main/java/com/gaethering/gaetheringserver/domain/board/service/CommentServiceImpl.java [hunk 6]
```diff
@@ -47,4 +50,34 @@ public CommentResponse writeComment(String email, Long postId, CommentRequest re
...
                 .build();
     }
+    @Override
+    @Transactional
+    public CommentResponse updateComment (String email, Long postId, Long commentId,
+                                          CommentRequest request) {
+
+        boolean result = postRepository.existsById(postId);
+
+        if(!result) {
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/service/CommentServiceImpl.java:61]: result가 여기서 한번만 사용이 된다면 if(!postRepository.existsById(postId)) 같은 방식으로 하는게 어떨까요?

File: src/main/java/com/gaethering/gaetheringserver/domain/board/service/CommentServiceImpl.java [hunk 7]
```diff
@@ -47,4 +50,34 @@ public CommentResponse writeComment(String email, Long postId, CommentRequest re
...
+                .orElseThrow(()-> new CommentNotFoundException());
+
+        Member member = memberRepository.findByEmail(email)
+                .orElseThrow(() -> new MemberNotFoundException());
+
+        if(!Objects.equals(member, comment.getMember())) {
+            throw new FailUpdateCommentException();
+        }
+        comment.setComment(request.getComment());
+        commentRepository.save(comment);
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/service/CommentServiceImpl.java:75]: 영속성 컨텍스트로 관리되어서 commentRepository.save(comment); 를 호출하지 않아도 될 것 같습니다!

File: src/main/java/com/gaethering/gaetheringserver/domain/board/controller/CommentController.java [hunk 8]
```diff
@@ -29,4 +29,18 @@ public ResponseEntity<CommentResponse> writeComment (Principal principal,
...
+    public ResponseEntity<CommentResponse> updateComment(Principal principal,
+                                                         @PathVariable Long postId,
+                                                         @PathVariable Long commentId,
+                                                         @RequestBody CommentRequest request) {
+
+        CommentResponse response =
+                commentService.updateComment(principal.getName(), postId, commentId, request);
+
+        return ResponseEntity.status(HttpStatus.OK)
+                .contentType(MediaType.APPLICATION_JSON).body(response);
```
Comment (younghoondoodoom) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/controller/CommentController.java:43]: 위와 같은 방식으로 하면 contentType을 굳이 APPLICATION_JSON으로 지정하지 않아도 자동을 될 것 같습니다!

**Review by younghoondoodoom — COMMENTED** [review]
주말에도 고생 많으셨습니다! 미처 놓치고 있던 부분도 코드 보면서 다시 생각이 나서 좋았습니다! 제 생각 코멘트 남겼으니 참고 부탁드립니다

File: src/main/java/com/gaethering/gaetheringserver/domain/board/entity/Comment.java [hunk 9]
```diff
@@ -10,24 +10,22 @@
...
 public class Comment extends BaseTimeEntity {
 
     @Id
     @GeneratedValue(strategy = GenerationType.IDENTITY)
     @Column(name = "comment_id", nullable = false)
     private Long id;
 
-    @Column(nullable = false)
+    @Column(columnDefinition = "TEXT", nullable = false)
     private String comment;
```
Comment (naekang) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/entity/Comment.java:29]: 댓글은 post와 다르게 길이 제한을 확실히 두는 것도 나쁘지 않을거같습니다!

File: src/main/java/com/gaethering/gaetheringserver/domain/board/exception/errorCode/PostErrorCode.java [hunk 10]
```diff
@@ -10,7 +10,10 @@ public enum PostErrorCode implements ErrorCode {
 
     POST_NOT_EXIST("E801", "존재하지 않는 게시물입니다."),
```
Comment (naekang) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/exception/errorCode/PostErrorCode.java:11]: 예외 클래스명과 맞게 `POST_NOT_FOUND`가 통일성 있어 보입니다!

File: src/main/java/com/gaethering/gaetheringserver/domain/board/exception/errorCode/PostErrorCode.java [hunk 11]
```diff
@@ -10,7 +10,10 @@ public enum PostErrorCode implements ErrorCode {
...
     CATEGORY_NOT_EXIST("E802", "존재하지 않는 카테고리입니다."),
```
Comment (naekang) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/exception/errorCode/PostErrorCode.java:12]: 위에 `POST`와 동일한 의견입니다 ㅎㅅㅎ

File: src/main/java/com/gaethering/gaetheringserver/domain/board/service/CommentServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+                .build();
+    }
+
+    @Override
+    @Transactional
+    public boolean deleteComment (String email, Long postId, Long commentId) {
+
+        boolean result = postRepository.existsById(postId);
+
+        if(!result) {
```
Comment (naekang) [inline src/main/java/com/gaethering/gaetheringserver/domain/board/service/CommentServiceImpl.java:90]: 이부분도 위의 `update`에서의 영훈님 의견과 동일합니다!

**Review by naekang — COMMENTED** [review]
주말동안 고생 많으셨습니다! 많이 배워 갑니다 ㅎㅎ 🤟

**Comment by naekang** [comment]
|File|Coverage [77.74%]|:x:|
|:-|:-:|:-:|
|[CommentController.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fcontroller%2FCommentController.java)|100%|:green_apple:|
|[MemberProfileServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fmember%2Fservice%2Fmember%2FMemberProfileServiceImpl.java)|100%|:green_apple:|
|[CommentServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fservice%2FCommentServiceImpl.java)|100%|:green_apple:|
|[FollowServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fmember%2Fservice%2Ffollow%2FFollowServiceImpl.java)|100%|:green_apple:|
|[PostErrorCode.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fexception%2FerrorCode%2FPostErrorCode.java)|100%|:green_apple:|
|[PetServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fpet%2Fservice%2FPetServiceImpl.java)|100%|:green_apple:|
|[MemberServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fmember%2Fservice%2Fmember%2FMemberServiceImpl.java)|87.15%|:green_apple:|
|[PostServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fservice%2FPostServiceImpl.java)|82.47%|:green_apple:|
|[S3ServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Faws%2Fs3%2FS3ServiceImpl.java)|80.62%|:green_apple:|
|[Category.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fentity%2FCategory.java)|74.32%|:x:|
|[Comment.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fentity%2FComment.java)|73.63%|:x:|
|[Post.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fentity%2FPost.java)|68.02%|:x:|
|[EmailServiceImpl.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Femail%2FEmailServiceImpl.java)|17.81%|:x:|
|[Heart.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fentity%2FHeart.java)|4.17%|:x:|
|[PostImage.java](https://github.com/Gaethering/Gaethering-Server/blob/6feb61d15d1433b2a01103b9734af96e327c5baf/src%2Fmain%2Fjava%2Fcom%2Fgaethering%2Fgaetheringserver%2Fdomain%2Fboard%2Fentity%2FPostImage.java)|3.61%|:x:|

|Total Project Coverage|64.18%|:x:|
|:-|:-:|:-:|

**Pull request merged**